internal/speakerlist/speakerlist.go

```diff
@@ -0,0 +1,215 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package speakerlist
+
+import (
+	"crypto/sha256"
+	"errors"
+	golog "log"
+	"reflect"
+	"strconv"
+	"sync"
+	"time"
+
+	"go.universe.tf/metallb/internal/k8s"
+	v1 "k8s.io/api/core/v1"
+
+	gokitlog "github.com/go-kit/kit/log"
+	"github.com/hashicorp/memberlist"
+)
+
+// SpeakerList gives you the list of healthy speaker
+type SpeakerList struct {
+	sync.Mutex
+	l           gokitlog.Logger
+	client      *k8s.Client
+	mList       *memberlist.Memberlist
+	k8sSpeakers map[string]string
+	eventCh     chan memberlist.NodeEvent
+	stopCh      chan struct{}
+}
+
+// New creates a new SpeakerList
+func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret string, stopCh chan struct{}) (*SpeakerList, error) {
+	sl := SpeakerList{
+		l:           logger,
+		k8sSpeakers: map[string]string{},
+		stopCh:      stopCh,
+	}
+
+	if bindAddr == "" {
+		logger.Log("op", "startup", "msg", "Not starting fast dead node detection (MemberList), need ml-bindaddr config")
+		return &sl, nil
+	}
+
+	mconfig := memberlist.DefaultLANConfig()
+	// mconfig.Name MUST be spec.nodeName, as we will match it against Enpoints nodeName in usableNodes()
+	mconfig.Name = nodeName
+	mconfig.BindAddr = bindAddr
+	if bindPort != "" {
+		mlport, err := strconv.Atoi(bindPort)
+		if err != nil {
+			sl.l.Log("op", "startup", "error", "unable to parse ml-bindport", "msg", err)
+			return nil, err
+		}
+		mconfig.BindPort = mlport
+		mconfig.AdvertisePort = mlport
+	}
+	loggerout := gokitlog.NewStdlibAdapter(gokitlog.With(sl.l, "component", "MemberList"))
+	mconfig.Logger = golog.New(loggerout, "", golog.Lshortfile)
+	if secret != "" {
+		sha := sha256.New()
+		mconfig.SecretKey = sha.Sum([]byte(secret))[:16]
+	}
+	sl.eventCh = make(chan memberlist.NodeEvent, 16)
+	mconfig.Events = &memberlist.ChannelEventDelegate{Ch: sl.eventCh}
+	var err error
+	sl.mList, err = memberlist.Create(mconfig)
+	if err != nil {
+		sl.l.Log("op", "startup", "error", err, "msg", "failed to create memberlist")
+		return nil, err
+	}
+	return &sl, nil
+}
+
+// Start starts the needed goroutines
+func (sl *SpeakerList) Start(client *k8s.Client) {
+	sl.client = client
+
+	if sl.mList == nil {
+		return
+	}
+
+	go sl.watchEvents()
+	go sl.memberlistJoinLoop()
+}
+
+func (sl *SpeakerList) memberlistJoin() {
+	if sl.mList == nil {
+		return
+	}
+	joinList := sl.joinList()
+	if len(joinList) > 0 {
```

Comment:
To keep code simpler, I prefer something like:

```suggestion
	if len(joinList) == 0 {
	    return
	}
```

And then just keep the code, but with one less level of identation.

I find it more readable, as it is perfectly clear the "uninteresting" cases (nothing to join, mList is nil, etc.) are filtered at the start and just return

Reply:
done